plugin/action/throttle/redis_limiter.go

```diff
@@ -0,0 +1,171 @@
+package throttle
+
+import (
+	"bytes"
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/ozontech/file.d/logger"
+	"github.com/ozontech/file.d/pipeline"
+)
+
+const (
+	keySuffix = "limit"
+)
+
+type redisLimiter struct {
+	mu    sync.Mutex
+	redis redisClient
+
+	// <keyPrefix>_
+	// will be used for bucket counter <keyPrefix>_<bucketID> and limit key <keyPrefix>_limit
+	keyPrefix bytes.Buffer
+
+	// contains values which will be used for incrementing remote bucket counter
+	// buckets will be flushed after every sync to contain only increment value
+	incrementLimiter *inMemoryLimiter
+
+	// contains global values synced from redis
+	totalLimiter *inMemoryLimiter
+}
+
+func NewRedisLimiter(
+	ctx context.Context,
+	redis redisClient,
+	throttleFieldName, throttleFieldValue string,
+	refreshInterval, bucketInterval time.Duration,
+	bucketCount int,
+	limit complexLimit,
+) *redisLimiter {
+	rl := &redisLimiter{
+		redis:            redis,
+		incrementLimiter: NewInMemoryLimiter(bucketInterval, bucketCount, limit),
+		totalLimiter:     NewInMemoryLimiter(bucketInterval, bucketCount, limit),
+	}
+
+	rl.keyPrefix = bytes.Buffer{}
+
+	// full name of keyPrefix will be $throttleFieldName_$throttleFieldValue_limit. `limit` added afterwards
+	rl.keyPrefix.WriteString(throttleFieldName)
+	rl.keyPrefix.WriteString("_")
+	rl.keyPrefix.WriteString(throttleFieldValue)
+	rl.keyPrefix.WriteString("_")
+
+	ticker := time.NewTicker(refreshInterval)
+	go func() {
```

Comment:
wow, it will spawn goroutine for each limit key, let's have a limited pool of global sync workers

Reply:
done